File: packages/connector/src/services/toggl-track/sync-masters.ts
/**
 * Toggl Track Master Data Sync
 *
 * Syncs projects, clients, tags, me, workspaces, users, groups.
 * All masters are fetched and saved in parallel.
 */

import { upsertRaw, RawRecord } from "../../db/raw-client.js";
import { setupLogger } from "../../lib/logger.js";
import {
  fetchProjects,
  fetchClients,
  fetchTags,
  fetchMe,
  fetchWorkspaces,
  fetchWorkspaceUsers,
  fetchWorkspaceGroups,
  getAuthInfo,
} from "./api-client.js";

const logger = setupLogger("toggl-masters");

const API_VERSION = "v9";

// Types
export interface MasterSyncResult {
  success: boolean;
  counts: Record<string, number>;
  elapsedSeconds: number;
}

/**
 * Convert API response to RawRecord list
 */
function toRawRecords(
  items: Record<string, unknown>[],
  idField: string = "id"
): RawRecord[] {
  return items.map((item) => ({
    sourceId: String(item[idField]),
    data: item,
  }));
}

async function syncProjects(): Promise<number> {
  const data = await fetchProjects();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__projects", records, API_VERSION);
  return result.total;
}

async function syncClients(): Promise<number> {
  const data = await fetchClients();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__clients", records, API_VERSION);
  return result.total;
}

async function syncTags(): Promise<number> {
  const data = await fetchTags();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__tags", records, API_VERSION);
  return result.total;
}

async function syncMe(): Promise<number> {
  const data = await fetchMe();
  if (!data || !data.id) return 0;
  const record: RawRecord = { sourceId: String(data.id), data };
  const result = await upsertRaw("toggl_track__me", [record], API_VERSION);
  return result.total;
}

async function syncWorkspaces(): Promise<number> {
  const data = await fetchWorkspaces();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__workspaces", records, API_VERSION);
  return result.total;
}

async function syncUsers(): Promise<number> {
  const data = await fetchWorkspaceUsers();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__users", records, API_VERSION);
  return result.total;
}

async function syncGroups(): Promise<number> {
  const data = await fetchWorkspaceGroups();
  if (data.length === 0) return 0;
  const records = toRawRecords(data);
  const result = await upsertRaw("toggl_track__groups", records, API_VERSION);
  return result.total;
}

/**
 * Sync all master data in parallel
 */
export async function syncMasters(): Promise<MasterSyncResult> {
  const startTime = performance.now();
  logger.info("Starting Toggl masters sync");

  try {
    // Warm up auth cache first to prevent parallel DB connections
    await getAuthInfo();

    // Fetch and save all masters in parallel
    const results = await Promise.allSettled([
      syncProjects(),
      syncClients(),
      syncTags(),
      syncMe(),
      syncWorkspaces(),
      syncUsers(),
      syncGroups(),
    ]);

    const masterNames = [
      "projects",
      "clients",
      "tags",
      "me",
      "workspaces",
      "users",
      "groups",
    ];
    const counts: Record<string, number> = {};
    const errors: string[] = [];

    for (let i = 0; i < masterNames.length; i++) {
      const name = masterNames[i];
      const result = results[i];

      if (result.status === "rejected") {
        logger.error(`Failed to sync ${name}: ${result.reason}`);
        errors.push(`${name}: ${result.reason}`);
        counts[name] = 0;
      } else {
        counts[name] = result.value;
      }
    }

    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;

    // Log summary
    const countsStr = Object.entries(counts)
      .map(([k, v]) => `${k}=${v}`)
      .join(", ");
    logger.info(`Toggl masters sync completed in ${elapsed}s: ${countsStr}`);

    if (errors.length > 0) {
      logger.warn(`Some masters failed: ${errors.join(", ")}`);
    }

    return {
      success: errors.length === 0,
      counts,
      elapsedSeconds: elapsed,
    };
  } catch (error) {
    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
    logger.error(`Toggl masters sync failed after ${elapsed}s: ${error}`);
    throw error;
  }
}

File: packages/connector/services/toggl_track/sync_masters.py
"""Toggl Track マスタデータ同期

projects, clients, tags, me, workspaces, users, groups を同期。
全マスタを並列で取得・保存。
"""

import asyncio
import time
from typing import Any, TypedDict

from db.raw_client import upsert_raw, RawRecord
from lib.logger import setup_logger
from services.toggl_track.api_client import (
    fetch_projects,
    fetch_clients,
    fetch_tags,
    fetch_me,
    fetch_workspaces,
    fetch_workspace_users,
    fetch_workspace_groups,
)

logger = setup_logger(__name__)

API_VERSION = "v9"


class MasterSyncResult(TypedDict):
    """マスタ同期結果"""
    success: bool
    counts: dict[str, int]
    elapsed_seconds: float


def _to_raw_records(items: list[dict[str, Any]], id_field: str = "id") -> list[RawRecord]:
    """APIレスポンスをRawRecordリストに変換"""
    return [
        RawRecord(source_id=str(item[id_field]), data=item)
        for item in items
    ]


async def _sync_projects() -> int:
    """プロジェクトを同期"""
    data = await fetch_projects()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__projects", records, api_version=API_VERSION)
    return result["total"]


async def _sync_clients() -> int:
    """クライアントを同期"""
    data = await fetch_clients()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__clients", records, api_version=API_VERSION)
    return result["total"]


async def _sync_tags() -> int:
    """タグを同期"""
    data = await fetch_tags()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__tags", records, api_version=API_VERSION)
    return result["total"]


async def _sync_me() -> int:
    """現在のユーザー情報を同期"""
    data = await fetch_me()
    if not data:
        return 0
    record = RawRecord(source_id=str(data["id"]), data=data)
    result = await upsert_raw("toggl_track__me", [record], api_version=API_VERSION)
    return result["total"]


async def _sync_workspaces() -> int:
    """ワークスペースを同期"""
    data = await fetch_workspaces()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__workspaces", records, api_version=API_VERSION)
    return result["total"]


async def _sync_users() -> int:
    """ワークスペースユーザーを同期"""
    data = await fetch_workspace_users()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__users", records, api_version=API_VERSION)
    return result["total"]


async def _sync_groups() -> int:
    """ワークスペースグループを同期"""
    data = await fetch_workspace_groups()
    if not data:
        return 0
    records = _to_raw_records(data)
    result = await upsert_raw("toggl_track__groups", records, api_version=API_VERSION)
    return result["total"]


async def sync_masters() -> MasterSyncResult:
    """全マスタデータを並列同期

    Returns:
        同期結果（各マスタの件数）
    """
    start_time = time.perf_counter()
    logger.info("Starting Toggl masters sync")

    try:
        # 全マスタを並列取得・保存
        results = await asyncio.gather(
            _sync_projects(),
            _sync_clients(),
            _sync_tags(),
            _sync_me(),
            _sync_workspaces(),
            _sync_users(),
            _sync_groups(),
            return_exceptions=True,
        )

        # 結果を整理
        master_names = ["projects", "clients", "tags", "me", "workspaces", "users", "groups"]
        counts = {}
        errors = []

        for name, result in zip(master_names, results):
            if isinstance(result, Exception):
                logger.error(f"Failed to sync {name}: {result}")
                errors.append(f"{name}: {result}")
                counts[name] = 0
            else:
                counts[name] = result

        elapsed = round(time.perf_counter() - start_time, 2)

        # ログ出力
        counts_str = ", ".join(f"{k}={v}" for k, v in counts.items())
        logger.info(f"Toggl masters sync completed in {elapsed}s: {counts_str}")

        if errors:
            logger.warning(f"Some masters failed: {errors}")

        return MasterSyncResult(
            success=len(errors) == 0,
            counts=counts,
            elapsed_seconds=elapsed,
        )

    except Exception as e:
        elapsed = round(time.perf_counter() - start_time, 2)
        logger.error(f"Toggl masters sync failed after {elapsed}s: {e}")
        raise
